### DISCUSSION TITLE
Exceptions Usage Help

### DISCUSSION THREAD

**Discussion by m4l490n** [nlohmann/json#2276] [Q&A] [answered]
Hello

I was wondering if it is possible to know more details when an exception is thrown. I can see [here](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_ac00185487a16455dfdd53884f678040d.html#ac00185487a16455dfdd53884f678040d) that only a "message" and "exception id" is being shown.

In my case, I'm having this exception:

`message: [json.exception.type_error.302] type must be number, but is string`

On a 47 element JSON. So having just that said make it difficult to know what element is the one with the string instead of a number"

**Comment by nlohmann** [comment]
I'm afraid there is no more information. Was there no stacktrace?

**Reply by m4l490n** [reply]
No stack trace, this is all I'm getting:

```
message: [json.exception.type_error.302] type must be string, but is number
exception id: 302
```

**Reply by nlohmann** [reply]
You get an exception without stacktrace? That is odd.

In order to asses your issue, we need the following information:

- Please describe the steps to reproduce the issue. Can you provide a small but working code example?

- What is the expected behavior?

- And what is the actual behavior instead?

- Which compiler and operating system are you using? Is it a [supported compiler](https://github.com/nlohmann/json#supported-compilers)?

- Did you use a released version of the library or the version from the `develop` branch?

- If you experience a compilation error: can you [compile and run the unit tests](https://github.com/nlohmann/json#execute-unit-tests)?

**Reply by nlohmann** [reply]
?

**Comment by psalvaggio** [comment]
For this reason, I normally use the following wrappers around this library:
```c++
struct field_type_error : public nlohmann::json::exception {
  field_type_error(const nlohmann::json::type_error& e, const char* field)
      : nlohmann::json::exception(e.id, (std::string("Error parsing field \"") +
                                         field + "\": " + e.what())
                                            .c_str()) {}
};

/** Parse a required field out of a JSON */
template <typename T>
void Required(const nlohmann::json& j, const char* key, T& v) try {
  j.at(key).get_to(v);
} catch (const nlohmann::json::type_error& e) {
  throw field_type_error(e, key);
}

/** \overload */
template <typename T>
void Required(const nlohmann::json& j, const std::string& key, T& v) try {
  j.at(key).get_to(v);
} catch (const nlohmann::json::type_error& e) {
  throw field_type_error(e, key.c_str());
}

/** Parse an optional field out of a JSON */
template <typename T>
void Optional(const nlohmann::json& j, const char* key, T& v) try {
  auto it = j.find(key);
  if (it != j.end()) {
    it.value().get_to(v);
  }
} catch (const nlohmann::json::type_error& e) {
  throw field_type_error(e, key);
}

/** \overload */
template <typename T>
void Optional(const nlohmann::json& j, const std::string& key, T& v) {
  Optional(j, key.c_str(), v);
}
```
and then in your `from_json()`, instead of using:
```c++
j.at("field").get_to(value);
```
use
```c++
Required(j, "field", value);
```
and you will get a more descriptive exception on a type error. It would be cool if something like this got included into the library, as type errors normally produce exceptions messages that are very hard to debug to the user.

**Reply by nlohmann** [reply]
Cool, I will have a look!

**Reply by nlohmann** [reply]
Hm. I had a look and tried with the following code:

```cpp
json j;
j["foo"] = 1;

int value;
Required(j, "field", value);
```

First, I had to adjust the exceptions, because internally, because `at` throws `json::out_of_range` instead of `json::type_error`.

Then, I get the following exception:

```
field_type_error: Error parsing field "field": [json.exception.out_of_range.403] key 'field' not found
```

whereas the original code

```cpp
json j;
j["foo"] = 1;

int value;
j.at("field").get_to(value);
```

throws

```
[json.exception.out_of_range.403] key 'field' not found
```

Am I using the code wrong? What is the benefit?

**Reply by psalvaggio** [reply]
So those wrappers are only trying to improve the errors on `json::type_error`. All of the other exceptions are passed through as is. Here's a simple test program:

```c++
#include <iostream>
  
#include <nlohmann/json.hpp>

struct field_type_error : public nlohmann::json::exception {
  field_type_error(const nlohmann::json::type_error& e, const char* field)
      : nlohmann::json::exception(e.id, (std::string("Error parsing field \"") +
                                         field + "\": " + e.what())
                                            .c_str()) {}
};

/** Parse a required field out of a JSON */
template <typename T>
void Required(const nlohmann::json& j, const char* key, T& v) try {
  j.at(key).get_to(v);
} catch (const nlohmann::json::type_error& e) {
  throw field_type_error(e, key);
}

int main(int argc, char** argv) {
  nlohmann::json j;
  j["foo"] = "bar";

  double a;
  try {
    j.at("foo").get_to(a);
  } catch (const std::exception& e) {
    std::cerr << e.what() << std::endl;
  }
  try {
    Required(j, "foo", a);
  } catch (const std::exception& e) {
    std::cerr << e.what() << std::endl;
  }

  return 0;
}
```
The output of this is:
```
[json.exception.type_error.302] type must be number, but is string
Error parsing field "foo": [json.exception.type_error.302] type must be number, but is string
```

**Reply by nlohmann** [reply]
Thanks, now I understand better.

**Reply by nlohmann** [reply]
FYI: This should be fixed with #2562 now. See https://json.nlohmann.me/home/exceptions/#extended-diagnostic-messages

**Reply by psalvaggio** [reply]
That's great! I'll definitely be converting my code to use this. Thanks for the feature!

**Comment by nlohmann** [comment, marked as answer]
This should be fixed with #2562 now. See https://json.nlohmann.me/home/exceptions/#extended-diagnostic-messages
